src/login/msal/MsalAuthProvider.ts

```diff
@@ -0,0 +1,122 @@
+		const authResult: AuthenticationResult | null = await this.publicClientApp.acquireTokenByCode({
+			scopes: msalScopes,
+			code,
+			redirectUri: redirectUrl,
+		});
+
+		if (!authResult) {
+			throw new Error(localize('azure-account.msalAuthFailed', 'MSAL authentication failed.'));
+		}
+
+		return authResult;
+	}
+
+	public async loginWithDeviceCode(): Promise<AbstractLoginResult> {
+		throw new Error('"Login With Device Code" not implemented for MSAL.');
+	}
+
+	public async loginSilent(): Promise<AbstractLoginResult> {
+		const msalTokenCache: TokenCache = this.publicClientApp.getTokenCache();
+		const accountInfo: AccountInfo[] = await msalTokenCache.getAllAccounts();
+		let authResult: AuthenticationResult | null;
+
+		if (accountInfo.length === 1) {
+			authResult = await this.publicClientApp.acquireTokenSilent({
+				scopes: msalScopes,
+				account: accountInfo[0]
+			});
+		} else {
+			throw new Error(localize('azure-account.expectedSingleAccount', 'Expected a single account when reading cache but multiple were found.'));
+		}
+
+		if (!authResult) {
+			throw new AzureLoginError(localize('azure-account.loginSilentFailed', 'Silent login failed.'));
+		}
+
+		return authResult;
+	}
+
+	public getCredentials(environment: string, userId: string, tenantId: string): DeviceTokenCredentials {
+		// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-explicit-any
+		return new DeviceTokenCredentials({ environment: (<any>Environment)[environment], username: userId, clientId, tokenCache: this.dummyTokenCache, domain: tenantId });
+	}
+
+	public getCredentials2(_env: Environment, _userId: string, _tenantId: string, accountInfo?: AccountInfo): AbstractCredentials2 {
+		// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+		return new AzureIdentityCredentialAdapter(new PublicClientCredential(this.publicClientApp, accountInfo!));
+	}
+
+	public async updateSessions(environment: Environment, loginResult: AbstractLoginResult, sessions: AzureSession[]): Promise<void> {
+		if (isAdalLoginResult(loginResult)) {
+			throw loginResultTypeError;
+		}
+
+		loginResult = <AuthenticationResult>loginResult;
+
+		/* eslint-disable @typescript-eslint/no-non-null-assertion */
+		sessions.splice(0, sessions.length, <AzureSession>{
+			environment,
+			userId: loginResult.account!.username,
+			tenantId: loginResult.account!.tenantId,
+			accountInfo: loginResult.account!,
+			// eslint-disable-next-line @typescript-eslint/no-explicit-any
+			credentials: this.getCredentials(<any>environment, loginResult.account!.username, loginResult.tenantId),
+			credentials2: this.getCredentials2(environment, loginResult.account!.username, loginResult.tenantId, loginResult.account!)
+		});
+		/* eslint-enable @typescript-eslint/no-non-null-assertion */
+	}
+
+	public async clearLibraryTokenCache(): Promise<void> {
+		// MSAL handles token caching under the hood
```

Comment:
They _can_ handle stuff under the hood for you, but you have to let them. Specifically, I think you need to call `TokenCache.removeAccount`. I believe that would be more robust than the stuff you're currently doing in `clearLocalTokenCache`, which should likely only be used for ADAL (aka just revert most of your changes that moved `AdalAuthProvider.clearTokenCache` into `AuthProviderBase.clearLocalTokenCache`).